File: py-sdk/src/forta_bot/cache/disk_cache.py
from os import path
from typing import Callable
from .cache import Cache


class DiskCache(Cache):
    def __init__(self, pickledb_load: Callable, folder_path: str):
        self.pickledb = pickledb_load(
            path.join(folder_path, "forta-bot-cache-py"), False)

    async def get_block_with_transactions(self, chain_id: int, block_hash_or_number: int | str) -> dict | None:
        return self.pickledb.get(self.get_block_with_transactions_key(chain_id, block_hash_or_number))

    async def set_block_with_transactions(self, chain_id: int, block: dict):
        # index by block hash
        self.pickledb.set(self.get_block_with_transactions_key(
            chain_id, block["hash"]), block)
        # also index by block number
        self.pickledb.set(self.get_block_with_transactions_key(
            chain_id, block["number"]), block)

    def get_block_with_transactions_key(self, chain_id: int, block_hash_or_number: int | str) -> str:
        return f'{chain_id}-{str(block_hash_or_number).lower()}'

    async def get_logs_for_block(self, chain_id: int, block_number: int) -> list[dict] | None:
        return self.pickledb.get(self.get_logs_for_block_key(chain_id, block_number))

    async def set_logs_for_block(self, chain_id: int, block_number: int, logs: list[dict]):
        self.pickledb.set(self.get_logs_for_block_key(
            chain_id, block_number), logs)

    def get_logs_for_block_key(self, chain_id: int, block_number: int) -> str:
        return f'{chain_id}-{block_number}-logs'

    async def get_trace_data(self, chain_id: int, block_number_or_tx_hash: int | str) -> list[dict] | None:
        return self.pickledb.get(self.get_trace_data_key(chain_id, block_number_or_tx_hash))

    async def set_trace_data(self, chain_id: int, block_number_or_tx_hash: int | str, traces: list[dict]):
        self.pickledb.set(self.get_trace_data_key(
            chain_id, block_number_or_tx_hash), traces)

    def get_trace_data_key(self, chain_id: int, block_number_or_tx_hash: int | str) -> str:
        return f'{chain_id}-{str(block_number_or_tx_hash).lower()}-trace'

    async def get_transaction_receipt(self, chain_id: int, tx_hash: str) -> dict | None:
        return self.pickledb.get(self.get_transaction_receipt_key(chain_id, tx_hash))

    async def set_transaction_receipt(self, chain_id: int, tx_hash: str, receipt: dict):
        self.pickledb.set(self.get_transaction_receipt_key(
            chain_id, tx_hash), receipt)

    def get_transaction_receipt_key(self, chain_id: int, tx_hash: str) -> str:
        return f'{chain_id}-{tx_hash.lower()}'

    async def get_alert(self, alert_hash: str) -> dict | None:
        return self.pickledb.get(self.get_alert_key(alert_hash))

    async def set_alert(self, alert_hash: str, alert: dict):
        self.pickledb.set(self.get_alert_key(alert_hash), alert)

    def get_alert_key(self, alert_hash: str) -> str:
        return f'{alert_hash.lower()}-alert'

    async def dump(self):
        self.pickledb.dump()  # writes to disk

File: ts-sdk/src/cache/disk.cache.ts
import { Cache as FlatCache } from "flat-cache";
import { JsonRpcBlock, JsonRpcLog, Trace } from "..";
import { Cache } from "./cache";
import { JsonRpcTransactionReceipt } from "../transactions";

export class DiskCache implements Cache {
  constructor(private readonly flatCache: FlatCache) {}

  async getLatestBlockNumber(chainId: number): Promise<string | undefined> {
    return undefined; // no-op
  }

  async getBlockWithTransactions(
    chainId: number,
    blockHashOrNumber: string | number
  ): Promise<JsonRpcBlock | undefined> {
    return this.flatCache.getKey(
      this.getBlockWithTransactionsKey(chainId, blockHashOrNumber)
    );
  }

  async setBlockWithTransactions(chainId: number, block: JsonRpcBlock) {
    // index by block hash
    this.flatCache.setKey(
      this.getBlockWithTransactionsKey(chainId, block.hash),
      block
    );
    // also index by block number
    this.flatCache.setKey(
      this.getBlockWithTransactionsKey(chainId, parseInt(block.number)),
      block
    );
  }

  private getBlockWithTransactionsKey(
    chainId: number,
    blockHashOrNumber: number | string
  ) {
    return `${chainId}-${blockHashOrNumber.toString().toLowerCase()}`;
  }

  async getLogsForBlock(
    chainId: number,
    blockNumber: number
  ): Promise<JsonRpcLog[] | undefined> {
    return this.flatCache.getKey(this.getLogsForBlockKey(chainId, blockNumber));
  }

  async setLogsForBlock(
    chainId: number,
    blockNumber: number,
    logs: JsonRpcLog[]
  ) {
    this.flatCache.setKey(this.getLogsForBlockKey(chainId, blockNumber), logs);
  }

  private getLogsForBlockKey(chainId: number, blockNumber: number) {
    return `${chainId}-${blockNumber}-logs`;
  }

  async getTraceData(
    chainId: number,
    blockNumberOrTxHash: string | number
  ): Promise<Trace[] | undefined> {
    return this.flatCache.getKey(
      this.getTraceDataKey(chainId, blockNumberOrTxHash)
    );
  }

  async setTraceData(
    chainId: number,
    blockNumberOrTxHash: string | number,
    traces: Trace[]
  ): Promise<void> {
    this.flatCache.setKey(
      this.getTraceDataKey(chainId, blockNumberOrTxHash),
      traces
    );
  }

  private getTraceDataKey(
    chainId: number,
    blockNumberOrTxHash: number | string
  ) {
    return `${chainId}-${blockNumberOrTxHash.toString().toLowerCase()}-trace`;
  }

  async getTransactionReceipt(
    chainId: number,
    txHash: string
  ): Promise<JsonRpcTransactionReceipt | undefined> {
    return this.flatCache.getKey(
      this.getTransactionReceiptKey(chainId, txHash)
    );
  }

  async setTransactionReceipt(
    chainId: number,
    txHash: string,
    receipt: JsonRpcTransactionReceipt
  ) {
    this.flatCache.setKey(
      this.getTransactionReceiptKey(chainId, txHash),
      receipt
    );
  }

  private getTransactionReceiptKey(chainId: number, txHash: string) {
    return `${chainId}-${txHash.toLowerCase()}`;
  }

  async getAlert(alertHash: string): Promise<object | undefined> {
    return this.flatCache.getKey(this.getAlertKey(alertHash));
  }

  async setAlert(alertHash: string, alert: object) {
    this.flatCache.setKey(this.getAlertKey(alertHash), alert);
  }

  private getAlertKey(alertHash: string) {
    return `${alertHash.toLowerCase()}-alert`;
  }

  async dump() {
    this.flatCache.save(true); // true = done remove keys not used in this run
  }
}
